### PR TITLE
Ct 2144 remove nplusone query issues

### PR DESCRIPTION
**Pull request by starswan**
Remove a lot of N+1 query issues (mainly on reports) to improve performance. Still one issue with Correspondence types

### CODE REVIEW

File: app/models/team.rb [hunk 1]
```diff
@@ -28,6 +28,11 @@ class Team < ApplicationRecord
   has_many :properties, class_name: TeamProperty, :dependent => :delete_all
```
Comment (mseedat-moj) [inline app/models/team.rb]: Can Properties also be eager loaded as that would ensure any calls to `team_lead` is not resource intensive

**Review by mseedat-moj — COMMENTED** [review]
Some initial comments. Generally very welcome changes

File: app/controllers/cases_controller.rb [hunk 2]
```diff
@@ -138,10 +142,16 @@ def my_open_cases
...
+    full_list_of_cases = @global_nav_manager
+                           .current_page_or_tab
+                           .cases
+                           .includes(:message_transitions,
+                                     :responder,
+                                     :approver_assignments,
+                                     :managing_team,
+                                     :responding_team)
+
+    query_list_params = filter_params.merge(list_path: request.path,)
```
Comment (mseedat-moj) [inline app/controllers/cases_controller.rb]: Forgot to delte comma

File: app/models/correspondence_type.rb [hunk 3]
```diff
@@ -68,6 +68,27 @@ class CorrespondenceType < ApplicationRecord
...
+    def find_by_abbrevation! abbreviation
+      all_cached.detect { |ct| ct.abbreviation == abbreviation } || super
+    end
+  end
+
+  # Invalidate the cache after saving a change
+  after_save do
+    self.class.clear_cache
+  end
+
```
Comment (mseedat-moj) [inline app/models/correspondence_type.rb]: Nice!

File: app/models/assignment.rb [hunk 4]
```diff
@@ -50,8 +51,10 @@ class Assignment < ApplicationRecord
 
   scope :pending_accepted, -> { where(state: %w[pending accepted]) }
 
+  # Putting a 'limit 1' here breaks the caching of Case::Base#responder_assignment
+  # please treat this as a 'private' method i.e. don't use it in application code
   scope :last_responding, -> {
-    responding.where.not(state: 'rejected').order(id: :desc).limit(1)
+    responding.where.not(state: 'rejected').order(id: :desc)
   }
```
[the remark this answers is not in the document]
Comment (starswan) [reply app/models/assignment.rb]: No that doesn't work, as it is needed by the case model.

File: spec/models/case/base_spec.rb [hunk 5]
```diff
@@ -1225,21 +1225,13 @@ def clear_case(kase, t, new_state, acting_team)
...
-      expect(kase.correspondence_type).to eq CorrespondenceType.foi
-    end
-
-    it 'only finds the correspondence_type once' do
-      foi = CorrespondenceType.foi
-      allow(CorrespondenceType).to receive(:find_by!).and_return(foi)
-      kase.correspondence_type
-      kase.correspondence_type
-      expect(CorrespondenceType).to have_received(:find_by!).at_least(1)
+      expect(kase.correspondence_type.abbreviation).to eq CorrespondenceType.foi.abbreviation
```
[the remark this answers is not in the document]
Comment (starswan) [reply spec/models/case/base_spec.rb]: They are no longer true (which is a good thing as it was creating an N+1 query issue all of its own)

File: app/models/team.rb (hunk 1 above)
Comment (starswan) [reply app/models/team.rb]: I've eager loaded team_leader which is all that is required.

File: spec/features/user_journeys/overturned_foi/ovt_foi_trigger_e2e_spec.rb [hunk 6]
```diff
@@ -27,7 +27,7 @@
   given!(:press_officer)        { create :press_officer }
   given(:press_office)          { press_officer.approving_team }
   given(:foi)                   { create :foi_correspondence_type }
-  given!(:private_officer)      { create :default_private_officer }
+  given!(:private_officer)      { find_or_create :default_private_officer }
```
[the remark this answers is not in the document]
Comment (starswan) [reply spec/features/user_journeys/overturned_foi/ovt_foi_trigger_e2e_spec.rb]: added comment explaining what I found. The patterns of find_or_create in this project are a little confusing - a combo of fixtures and FactoryBot

**Review by mseedat-moj — APPROVED** [review]

**Pull request merged**